From d46b243c6330cdabbfb7ef12e13fb5fdcd4e0161 Mon Sep 17 00:00:00 2001
From: Michael-CY Lee <michael-cy.lee@mediatek.com>
Date: Thu, 16 Oct 2025 11:02:00 +0800
Subject: [PATCH 009/105] wifi: mt76: reset channel state during scanning

Reset the channel state of all channels during scanning to prevent the
ACS algorithm from selecting unexpected channel.

Signed-off-by: Michael-CY Lee <michael-cy.lee@mediatek.com>
Signed-off-by: Shayne Chen <shayne.chen@mediatek.com>
---
 mac80211.c | 3 ++-
 scan.c     | 1 +
 2 files changed, 3 insertions(+), 1 deletion(-)

diff --git a/mac80211.c b/mac80211.c
index 8bb7403f8..c687fd6a9 100644
--- a/mac80211.c
+++ b/mac80211.c
@@ -1039,7 +1039,8 @@ int __mt76_set_channel(struct mt76_phy *phy, struct cfg80211_chan_def *chandef,
 	if (!offchannel)
 		phy->main_chandef = *chandef;
 
-	if (chandef->chan != phy->main_chandef.chan)
+	if (chandef->chan != phy->main_chandef.chan ||
+	    test_bit(MT76_SCANNING, &phy->state))
 		memset(phy->chan_state, 0, sizeof(*phy->chan_state));
 
 	ret = dev->drv->set_channel(phy);
diff --git a/scan.c b/scan.c
index ff9176cde..cf05e0f18 100644
--- a/scan.c
+++ b/scan.c
@@ -156,6 +156,7 @@ int mt76_hw_scan(struct ieee80211_hw *hw, struct ieee80211_vif *vif,
 	dev->scan.vif = vif;
 	dev->scan.phy = phy;
 	dev->scan.mlink = mlink;
+	set_bit(MT76_SCANNING, &phy->state);
 	ieee80211_queue_delayed_work(dev->phy.hw, &dev->scan_work, 0);
 
 out:
-- 
2.45.2

